[DataGridPremium] Reordering support for row grouping (#18251)
mui/mui-x: docs/data/data-grid/row-grouping/row-grouping.md

:::

## Reorder row grouping rows

Row reordering allows the users to drag and drop the rows to update their order.
To enable this feature with the row grouping, pass the `rowReordering` prop on the Data Grid Premium component:

```tsx
<DataGridPremium rowGroupingModel={['category']} rowReordering />
```

{{"demo": "RowGroupingReordering.js", "bg": "inline", "defaultCodeOpen": false}}

:::warning

> There's currently a limitation to the multi-level row grouping. You can not move a parent on second or higher to another parent at a lower level because it will essentially invoke bulk row editing.

Author: 👍 I updated it a bit further.

> For example in movies grouped by Company and Director:

Reviewer: ```suggestion
For example, in movies grouped by Company and Director:
```

> ⛔️ You can not move a director parent row from one company to another, because it would mean multiple rows being updated i.e. multiple `processRowUpdate` calls.

Reviewer: ```suggestion
⛔️ You can not move a director parent row from one company to another, because it would mean multiple rows being updated i.e. multiple `processRowUpdate()` calls.
```

✅ You can reorder leaf rows from one director to another or move directors around within same company.

Please open a new issue if that is a use-case you are interested in.

:::

### Usage with groupingValueSetter()

> If you use [`colDef.groupingValueGetter()`](#using-groupingvaluegetter-for-complex-grouping-value) to handle complex grouping values, in order for grouping across rows to work, you must also use the `colDef.groupingValueSetter()` which is essentially the inverse of `groupingValueGetter()`.

Reviewer: ```suggestion
If you use [`colDef.groupingValueGetter()`](#using-groupingvaluegetter-for-complex-grouping-value) to handle complex grouping values, you must use the `colDef.groupingValueSetter()` to properly set convert back the simple value back to the complex grouping value.
```

+ company: 'Disney Studios',
  ...
 }
```

The Data Grid updates the internal row data, but for the change to persist on the server, you must use the [`processRowUpdate()`](/x/react-data-grid/editing/persistence/#the-processrowupdate-callback) callback.

When moving a nested group to a different parent on the same depth, there will be multiple `processRowUpdate()` calls for all the leaf descendants of the dragged group. In case some of them fail, you might see the parent node appearing in both places. Consider using the `onProcessRowUpdateError()` to show the proper feedback to the user.

### Usage with groupingValueSetter()

If you use [`colDef.groupingValueGetter()`](#using-groupingvaluegetter-for-complex-grouping-value) to handle complex grouping values and you want to group across rows, you must use the `colDef.groupingValueSetter()` to properly convert back the simple value to the complex one.

This method should return the updated row based on the groupKey (`value`) that corresponds to the target group.

```ts
const columns: GridColDef[] = [
  {
    field: 'composer',
    groupingValueGetter: (value) => value.name,
    groupingValueSetter: (value, row) => ({
      ...row,

>       composer: { name: value },

Reviewer: Should this be
```suggestion
      composer: {
        ...row.composer,
        name: value,
      },
```
to keep other `composer` props?

Author: I didn't put it because `composer` in the `useMovieData` (where this data is coming from) has no other props.
https://github.com/mui/mui-x/blob/19988b53dbdf5bf8a8cbde55723dfa8bebdeccca/packages/x-data-grid-generator/src/hooks/useMovieData.ts#L12

Do you think it will still be useful to add a destructuring, may be to promote good practices?

:::

## Reorder row grouping rows

Row reordering allows the users to drag and drop the rows to update their order.
To enable this feature with the row grouping, pass the `rowReordering` prop on the Data Grid Premium component:

```tsx
<DataGridPremium rowGroupingModel={['category']} rowReordering />
```

> {{"demo": "RowGroupingReordering.js", "bg": "inline", "defaultCodeOpen": false}}

Reviewer 2: I noticed that if I move the row to a different group, and then stop grouping by company, the row data is reverted to the original.
Is this expected?

Author: >  Is this expected?

No, it was a bug. The row update was incomplete. Fixed: https://github.com/mui/mui-x/pull/18251/commits/fcf5d17c1bbbe7babc9ef285a53c954c56f2e90f

### Reacting to group updates

Wheneven a row is moved from one group to another, it warrants a row update, the row data value that was used to group this row must now be updated to maintain the row grouping data integrity.

For example, on a Data Grid having movies grouped by companies, if a row "Avatar" is moved from "20th Century Fox" to "Disney Studios", apart from the row being updated in the row tree, the row data must be updated to reflect this change.

```diff
 // "Avatar" row
 {
  title: 'Avatar',
- company: '20th Century Fox',
+ company: 'Disney Studios',
  ...
 }
```

> The Data Grid handles this row update automatically unless there's a [`processRowUpdate()`](/x/react-data-grid/editing/persistence/#the-processrowupdate-callback) callback passed in which case, the user can handle how the row will be processed.

Reviewer 2: ```suggestion
The Data Grid updates the row internally, but to persist the change you must use the [`processRowUpdate()`](/x/react-data-grid/editing/persistence/#the-processrowupdate-callback) callback.
```

Do I understand this correctly?

### Usage with groupingValueSetter()

If you use [`colDef.groupingValueGetter()`](#using-groupingvaluegetter-for-complex-grouping-value) to handle complex grouping values, in order for grouping across rows to work, you must use the `colDef.groupingValueSetter()` to properly convert back the simple value to the complex one.

It should return the updated row based on the groupKey (`value`) corresponding to the target group.

```ts
const columns: GridColDef[] = [
  {
    field: 'composer',
    groupingValueGetter: (value) => value.name,
    groupingValueSetter: (value, row) => ({
      ...row,
      composer: { name: value },
    }),
  },
  // ...
];
```

{{"demo": "RowGroupingGroupingValueSetter.js", "bg": "inline", "defaultCodeOpen": false}}

:::warning

There are some limitations when reordering grouped rows:

- Leaf rows (the lowest level) can only be moved within their current group or another group at the same level—they cannot become parents.

> - Parent rows can only be reordered among other parents at the same level; they cannot be moved to a different level or group.

Reviewer 2: Why do we set this limitation?
This operation is basically editing of multiple rows.

Author: @cherniavskii Thanks for linking the clipboard paste logic.

IIUC, clipboard paste is usually a **client-side** operation. Even when it’s backed by a server update, partial updates (i.e., some rows succeed, others fail) are acceptable by simply discarding the rows that failed.

However, row reordering seems to be a bit different. For example, if a grouped row “X” is moved from parent group **B** to **A**, and only 5 out of 10 child rows update successfully, the group could end up appearing in **both** groups. This leads to a confusing and inconsistent state.

Take this [demo](https://deploy-preview-18251--material-ui-x.netlify.app/x/react-data-grid/row-grouping/#initializing-row-grouping) as an example when moving “Anthony & Joe Russo” to “20th Century Fox”.
If some rows are updated and others aren’t, the group could appear under both “20th Century Fox” and “Disney Studios”.

As you mentioned, one option is to **revert the update entirely** if there’s a failure. But by that time, the backend will have already committed partial changes for the succeeded rows, and a revert may cause inconsistent client and server states.

So I’m curious — how would you feel about a scenario where, after a reorder, the group “Anthony & Joe Russo” ends up shown under both “20th Century Fox” and “Disney Studios”?
Would you consider that an acceptable edge case or a broken state from a user’s perspective?

Author: >From what I can tell, it's the same for clipboard paste. If the pasted data impacts 10 rows, processRowUpdate will be called 10 times (for each row individually), and the failed server updates will mean that these rows will not be changed.

@cherniavskii Thank you for your feedback. It made me realize we have to do it inevitably for the tree data feature so I did it for row grouping.

Partial movement of nested groups is supported now as we could have same director working for multiple companies. (May be part time or on project basis 😄)

I'd still like to change the signature of `processRowUpdate()` to handle multiple rows at a time in v9.

>This behavior looks good to me.
>I think we can assume 2 categories of server update failures here:
>
>    You don't have permission to update specific rows, or there's a bug in the backend and it throws error.
>    In this case, the row should stay in the original position.
>    Connectivity issue (network, server) – if you have a mechanism to retry failed requests – this will be transparent from the grid point of view. If not – the row will stay at the original positions.

I didn't classify the types of errors with the current PR, aiming to give it a shot with #18947, would be a nice behavior if we are able to handle connectivity issues.

:::

> ## Reorder grouped rows

Author: I added a compact version of it: https://github.com/mui/mui-x/pull/18251/commits/a97e02bd1f821e3b259cfdbb7ffe01658cd6525a

Lmk how it feels.

> ## Reorder row grouping rows
> 
> Row reordering allows the users to drag and drop the rows to update their order.
> To enable this feature with the row grouping, pass the `rowReordering` prop on the Data Grid Premium component:

Reviewer 2: ```suggestion
## Reorder grouped rows

Row reordering allows the users to reorder row groups, or move rows from one group to another.
To enable this feature with the row grouping, pass the `rowReordering` prop to the Data Grid Premium component:
```

```tsx
<DataGridPremium rowGroupingModel={['category']} rowReordering />
```

{{"demo": "RowGroupingReordering.js", "bg": "inline", "defaultCodeOpen": false}}

### Reacting to group updates

> Wheneven a row is moved from one group to another, it warrants a row update, the row data value that was used to group this row must now be updated to maintain the row grouping data integrity.

Reviewer 2: ```suggestion
Whenever a row is moved from one group to another, it warrants a row update, the row data value that was used to group this row must now be updated to maintain the row grouping data integrity.
```

> For example, on a Data Grid having movies grouped by companies, if a row "Avatar" is moved from "20th Century Fox" to "Disney Studios", apart from the row being updated in the row tree, the row data must be updated to reflect this change.

Reviewer 2: ```suggestion
For example, on a Data Grid having movies grouped by companies, if a row "Avatar" is moved from "20th Century Fox" to "Disney Studios" group, apart from the row being updated in the row tree, the row data must be updated to reflect this change.
```

```diff
 // "Avatar" row
 {
  title: 'Avatar',
- company: '20th Century Fox',
+ company: 'Disney Studios',
  ...
 }
```

> The Data Grid updates the row internally, to persist the change on the server [`processRowUpdate()`](/x/react-data-grid/editing/persistence/#the-processrowupdate-callback) callback must be used.

Reviewer 2: ```suggestion
The Data Grid updates the internal row data, but to persist the change on the server the [`processRowUpdate()`](/x/react-data-grid/editing/persistence/#the-processrowupdate-callback) callback must be used.
```

    groupingValueGetter: (value) => value.name,
    groupingValueSetter: (value, row) => ({
      ...row,
      composer: { name: value },
    }),
  },
  // ...
];
```

{{"demo": "RowGroupingGroupingValueSetter.js", "bg": "inline", "defaultCodeOpen": false}}

:::warning

There are some limitations when reordering grouped rows:

- Leaf rows (the lowest level) can only be moved within their current group or another group at the same level—they cannot become parents.
- Parent rows can only be reordered among other parents at the same level; they cannot be moved to a different level or group.

**For single-level grouping**: You can move leaf rows between any parent, and parents can be reordered like regular rows.

**For multi-level grouping**: Top-level groups can be reordered freely, but lower-level parents can only be reordered within their own parent group.

For example, in movies grouped by "Company" and "Director":

⛔️ You can not move a director parent row from one company to another, because it would mean multiple rows being updated i.e. multiple `processRowUpdate()` calls.

✅ You can reorder leaf rows from one director to another or move directors around within same company.

> Please open a [new issue](https://github.com/mui/mui-x/issues/new/choose) if that is a use-case you are interested in.

Author: I removed this, not relevant anymore.
